### ISSUE TITLE
add pairwise option to `adjsets` information

### ISSUE THREAD

**Issue by xjrc** [LLNL/conduit#777]
In order to better support a number of codes, an option should be added to `adjsests` instances to specify whether their `groups` members are "pairwise" or "maximally shared." Given that these two representations are functionally equivalent, Conduit/Blueprint should support transformations between these representations as well. The example below demonstrates both of these representations:

#### Diagram ####

```
Y
 
^ (0, 2)
|   +-----------++-----------+
|   |           ||           |
|   |           ||           |
|   |  domain2  ||  domain3  |
|   |           ||           |
|   |           ||           |
|   +-----------++-----------+
|   +-----------++-----------+
|   |           ||           |
|   |           ||           |
|   |  domain0  ||  domain1  |
|   |           ||           |
|   |           ||           |
|   +-----------++-----------+
|                          (2, 0)
+----------------------------->  X
```

#### YAML ####

```
domain0:
  coordsets:
    cset:
      type: "explicit"
      values:
        x: [0.0, 1.0, 0.0, 1.0]
        y: [0.0, 0.0, 1.0, 1.0]
  topologies:
    elem_topo:
      coordset: "cset"
      type: "unstructured"
      elements:
        shape: "quad"
        connectivity: [0, 1, 3, 2]
domain1:
  coordsets:
    coords:
      type: "explicit"
      values:
        x: [1.0, 2.0, 1.0, 2.0]
        y: [0.0, 0.0, 1.0, 1.0]
  topologies:
    elem_topo:
      coordset: "cset"
      type: "unstructured"
      elements:
        shape: "quad"
        connectivity: [0, 1, 3, 2]
domain2:
  coordsets:
    coords:
      type: "explicit"
      values:
        x: [0.0, 1.0, 0.0, 1.0]
        y: [1.0, 1.0, 2.0, 2.0]
  topologies:
    elem_topo:
      coordset: "cset"
      type: "unstructured"
      elements:
        shape: "quad"
        connectivity: [0, 1, 3, 2]
domain3:
  coordsets:
    coords:
      type: "explicit"
      values:
        x: [1.0, 2.0, 1.0, 2.0]
        y: [1.0, 1.0, 2.0, 2.0]
  topologies:
    elem_topo:
      coordset: "cset"
      type: "unstructured"
      elements:
        shape: "quad"
        connectivity: [0, 1, 3, 2]
```

### Maximally Shared Adjset (Current) ### 

```
domain0:
  ...
  adjsets:
    elem_aset:
      topology: "elem_topo"
      association: "vertex"
      groups:
        group01:
          neighbors: [1]
          values: [1]
        group02:
          neighbors: [2]
          values: [2]
        group0123:
          neighbors: [1, 2, 3]
          values: [3]
domain1:
  ...
  adjsets:
    elem_aset:
      topology: "elem_topo"
      association: "vertex"
      groups:
        group01:
          neighbors: [0]
          values: [0]
        group13:
          neighbors: [3]
          values: [3]
        group0123:
          neighbors: [1, 2, 3]
          values: [2]
domain2:
  ...
  adjsets:
    elem_aset:
      topology: "elem_topo"
      association: "vertex"
      groups:
        group02:
          neighbors: [0]
          values: [0]
        group23:
          neighbors: [3]
          values: [3]
        group0123:
          neighbors: [1, 2, 3]
          values: [1]
domain3:
  ...
  adjsets:
    elem_aset:
      topology: "elem_topo"
      association: "vertex"
      groups:
        group13:
          neighbors: [1]
          values: [1]
        group23:
          neighbors: [2]
          values: [2]
        group0123:
          neighbors: [0, 1, 2]
          values: [0]
```

###  Pairwise Adjset (Proposal) ###

```
domain0:
  ...
  adjsets:
    elem_aset:
      topology: "elem_topo"
      association: "vertex"
      groups:
        group01:
          neighbors: [1]
          values: [1, 3]
        group02:
          neighbors: [2]
          values: [2, 3]
        group03:
          neighbors: [3]
          values: [3]
domain1:
  ...
  adjsets:
    elem_aset:
      topology: "elem_topo"
      association: "vertex"
      groups:
        group01:
          neighbors: [0]
          values: [0, 2]
        group02:
          neighbors: [2]
          values: [2]
        group13:
          neighbors: [3]
          values: [2, 3]
domain2:
  ...
  adjsets:
    elem_aset:
      topology: "elem_topo"
      association: "vertex"
      groups:
        group02:
          neighbors: [0]
          values: [0, 1]
        group12:
          neighbors: [1]
          values: [1]
        group23:
          neighbors: [3]
          values: [1, 3]
domain3:
  ...
  adjsets:
    elem_aset:
      topology: "elem_topo"
      association: "vertex"
      groups:
        group03:
          neighbors: [0]
          values: [0]
        group13:
          neighbors: [1]
          values: [0, 1]
        group23:
          neighbors: [2]
          values: [0, 2]
```

Tags: @bryujin, @cyrush

**Comment by cyrush** [comment]
i believe the current blueprint is general enough to support both cases, so we should add properties to identify the cases and transforms to convert between the two

**Comment by xjrc** [comment]
@cyrush: Yes, agreed. I'd like to include this in #767 because I need it for some related processing. Here's a starter proposal:

## Spec Changes ##

```
  adjsets/adjset/association: "vertex" | "element"
  adjsets/adjset/topology: "topo"
+ adjsets/adjset/type: "pairwise" | "maxshare" | "unspecified" (default)
  adjsets/adjset/groups/group/neighbors: (integer array)
  adjsets/adjset/groups/group/values: (integer array)
```

I've included an `unspecified` category to provide room for codes that have unique representation needs.

We could also omit `type` and use data/tree structure to define these different styles instead (e.g. pairwise must have all `neighbors` arrays at length 1). This is similar in some ways to `matset`, but different in that this information cannot be determined in constant time (in either case, the array data in `groups` must be inspected).

## API Changes ##

```
+ bool conduit::blueprint::mesh::adjset::is_pairwise(const conduit::Node &n);
+ bool conduit::blueprint::mesh::adjset::is_maxshare(const conduit::Node &n);

+ bool conduit::blueprint::mesh::adjset::to_pairwise(const conduit::Node &src, conduit::Node &dest);
+ bool conduit::blueprint::mesh::adjset::to_maxshare(const conduit::Node &src, conduit::Node &dest);
```

**Comment by cyrush** [comment]
I vote for omitting the type as part of the what is presented and discovering it.

**Comment by xjrc** [comment]
I think that's ultimately my preference as well, though it does incur a O(# of groups) cost, which is unique for a check like this. Do you think that's a permissible deviation?

**Comment by xjrc** [comment]
(Perhaps already answered in your original response, but worth double checking.)

**Comment by cyrush** [comment]
two things to consider:  Whats the cost during `verify` and what is the cost during the `is_` routines.

I think the `is_` routines can be higher cost, but also open to higher costs for verify.

**Comment by xjrc** [comment]
All of the current `is_` routines (i.e. `is_multi_domain`, `is_multi_buffer`, `is_element_dominant`, etc.) are O(1) time, and all of the `verify` routines are O(# nodes in subtree) time. I think higher accuracy, higher cost `verify` routines are a good idea so long as this accuracy can be dialed in (e.g. checking data integrity instead of just types/structure), and I'm not opposed to making `is_*` checks as expensive as our existing `verify` routines.

I'll proceed with the type-less O(# groups) `is_*` solution; we can always reconsider our stance on the `is_*` functions and change them to a `verify` variant or something if we so choose later.

**Comment by xjrc** [comment]
After working on this a bit, I think the data check is definitely the way to go. Most reasonable definitions of "pair-wise" and "max-sharing" adjacency sets cause the two categories to be non-disjoint (consider an adjacency set for just two abutting domains; this can be considered both "pair-wise" and "max-sharing"), which makes the `type` classification untenable.

Regardless, getting firm definitions for each these terms is important. Here are some high-level rough drafts:

- **Pair-wise**: All groups have 1 neighbor and each neighbor set is unique.
- **Max-sharing**: All shared indices are unique and each neighbor set is unique.

In both of these cases, the first part of the clause is the simpler check, but it doesn't prevent an adjacency set from using 2+ groups to describe the data between equivalent neighbor sets (which seems like an option, based on [the documentation](https://llnl-conduit.readthedocs.io/en/latest/blueprint_mesh.html#adjacency-sets)). Perhaps we should make the second clause a part of the top-level adjacency set specification and include it in the future, more compute-intensive `adjset::verify` function?

**Comment by cyrush** [comment]
Do we see a use case to transform `max-shared` into pairwise? If not, we might not need to even identify `pair-wse`.

Restated: Regardless of what is presented, do we feel that the (potentially de-duplicted) "maximally shared" representation is the one we should target our transforms to?

If so we might just need `is_max_shared` and `to_max_shared` (but with better names)

I guess I view it as:

`maximally shared` and everything else, of which `pair-wise` is one more the more reasonable/common forms of everything else.

**Comment by xjrc** [comment]
Yes, we need `max-shared` to `pairwise` to go from MFEM (`max-shared`) to @terryhaut's code (`pairwise`). `is_pairwise` and `to_pairwise` are the key functions I need in the near-term; the `max-shared` equivalents are just bonuses (which we'll need to load external adjacency information into MFEM, but I don't know of a use case for this at present).

**Comment by cyrush** [comment]
Got it, thanks for filling that gap for me :-)

**Comment by xjrc** [comment]
I thought about naming a little bit more. How about instead of `max_shared`, we use the term `compact` or `compact_grouping` (to imply "we've grouped neighbors as much as possible")? These both may a bit overly restrictive as they don't allow for multiple groups spanning a single abutment, but this may be more inline with what MFEM requires/expects.

**Comment by cyrush** [comment]
We think this might be done.

**Comment by cyrush** [comment]
resolved in #767
